### PR TITLE
Many type improvements

### PR DESCRIPTION
**Pull request by bartv**
# Description

- Fixed many untyped methods
- Fixed many optionals
- The precision of all modules is now below 20%

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [x] Attached issue to pull request
- [x] Changelog entry
- [x] Type annotations are present
- [x] Code is clear and sufficiently documented
- [x] No (preventable) type errors (check using make mypy or make mypy-diff)
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design
- [x] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )

# Reviewer Checklist:

- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Code is clear and sufficiently documented
- [ ] Correct, in line with design

### CODE REVIEW

**Review by wouterdb — CHANGES_REQUESTED** [review]

File: src/inmanta/agent/handler.py [hunk 1]
```diff
@@ -620,7 +621,7 @@ def _should_reload() -> bool:
             if not self.can_reload():
                 return False
             result = self.run_sync(_call_resource_did_dependency_change)
-            if result.code != 200:
+            if not result.result or result.code != 200:
                 error_msg_from_server = f": {result.result['message']}" if "message" in result.result else ""
```
[the remark this answers is not in the document]
Comment (bartv) [reply src/inmanta/agent/handler.py:628]: good catch, fixed now

File: src/inmanta/data/__init__.py [hunk 2]
```diff
@@ -117,13 +117,13 @@ def __init__(self, total: int, before: int, after: int) -> None:
 
 
 class InvalidQueryParameter(Exception):
-    def __init__(self, message: str, *args) -> None:
```
[the remark this answers is not in the document]
Comment (bartv) [reply src/inmanta/data/__init__.py:120]: agreed

File: src/inmanta/main.py [hunk 3]
```diff
@@ -815,7 +819,7 @@ def resource_action_log(ctx: click.Context) -> None:
 
 
 def validate_resource_version_id(
-    ctx: click.Context, option: Union[click.Option, click.Parameter], value: Any
+    ctx: click.Context, option: Union[click.Option, click.Parameter], value: ResourceVersionIdStr
```
[the remark this answers is not in the document]
Comment (bartv) [reply src/inmanta/main.py]: Fixed

**Review by wouterdb — APPROVED** [review]

File: src/inmanta/server/agentmanager.py [hunk 4]
```diff
@@ -910,7 +911,7 @@ async def _ensure_agents(self, env: data.Environment, agents: List[str], restart
             raise ShutdownInProgress()
 
         agent_map: Dict[str, str]
-        agent_map = await env.get(data.AUTOSTART_AGENT_MAP)
+        agent_map = cast(Dict[str, str], await env.get(data.AUTOSTART_AGENT_MAP))  # we know the type of this map
```
Comment (sanderr) [inline src/inmanta/server/agentmanager.py:914]: Same comment here (and below) with regards to the cast.

File: src/inmanta/protocol/rest/client.py [hunk 5]
```diff
@@ -91,7 +91,9 @@ def _get_client_config(self) -> str:
 
         return "%s://%s:%d" % (protocol, host, port)
 
-    async def call(self, properties: common.MethodProperties, args: List, kwargs: Dict[str, Any] = None) -> common.Result:
+    async def call(
+        self, properties: common.MethodProperties, args: List, kwargs: Optional[Dict[str, Any]] = None
```
Comment (sanderr) [inline src/inmanta/protocol/rest/client.py:95]:
```suggestion
        self, properties: common.MethodProperties, args: List[object], kwargs: Optional[Dict[str, Any]] = None
```

**Review by sanderr — CHANGES_REQUESTED** [review]

File: src/inmanta/config.py [hunk 6]
```diff
@@ -101,6 +104,16 @@ def _get_instance(cls) -> ConfigParser:
     def _reset(cls) -> None:
         cls.__instance = None
 
+    @overload
+    @classmethod
+    def get(cls, section: Optional[str] = None) -> ConfigParser:
```
[the remark this answers is not in the document]
Comment (bartv) [reply src/inmanta/config.py:109]: no, you are right. I typed it like I thought it worked but I should have read the code better.

File: src/inmanta/config.py [hunk 7]
```diff
@@ -101,6 +104,16 @@ def _get_instance(cls) -> ConfigParser:
...
+        ...
+
+    @overload
+    @classmethod
+    def get(cls, section: str, name: str, default_value: Optional[str] = None) -> str:
```
[the remark this answers is not in the document]
Comment (bartv) [reply src/inmanta/config.py:114]: you are right

File: src/inmanta/data/__init__.py [hunk 8]
```diff
@@ -756,7 +760,9 @@ async def delete_all(cls, connection: Optional[asyncpg.connection.Connection] =
         return record_count
 
     @classmethod
-    def _get_composed_filter(cls, offset: int = 1, col_name_prefix: str = None, **query: Any) -> Tuple[str, List[Any]]:
+    def _get_composed_filter(
+        cls, offset: int = 1, col_name_prefix: Optional[str] = None, **query: Any
+    ) -> Tuple[str, List[Any]]:
```
[the remark this answers is not in the document]
Comment (bartv) [reply src/inmanta/data/__init__.py:765]: missed it

File: src/inmanta/execute/proxy.py [hunk 9]
```diff
@@ -146,7 +148,7 @@ def return_value(cls, value: object) -> Union[None, str, tuple, int, float, bool
 
         return DynamicProxy(value)
 
-    def __getattr__(self, attribute):
+    def __getattr__(self, attribute: str) -> "DynamicProxy":
```
[the remark this answers is not in the document]
Comment (bartv) [reply src/inmanta/execute/proxy.py:151]: you are right, I am going to remove it for know then

File: src/inmanta/server/agentmanager.py (hunk 4 above)
Comment (bartv) [reply src/inmanta/server/agentmanager.py:914]: you mean that I should add the comment in those places as well?

File: src/inmanta/util.py [hunk 10]
```diff
@@ -273,7 +273,7 @@ def _custom_json_encoder(o: object) -> Union[ReturnTypes, "JSONSerializable"]:
         return o.isoformat(timespec="microseconds")
 
     if hasattr(o, "to_dict"):
-        return o.to_dict()
+        return o.to_dict()  # type: ignore
```
[the remark this answers is not in the document]
Comment (sanderr) [reply src/inmanta/util.py:276]: Interesting, I didn't know about `typing_extensions`.

**Review by sanderr — APPROVED** [review]

File: src/inmanta/server/services/orchestrationservice.py [hunk 11]
```diff
@@ -389,8 +389,8 @@ def safe_get(input: JsonType, key: str, default: object) -> object:
         auto_deploy = await env.get(data.AUTO_DEPLOY)
         if auto_deploy:
             LOGGER.debug("Auto deploying version %d", version)
-            push_on_auto_deploy: bool = await env.get(data.PUSH_ON_AUTO_DEPLOY)
-            agent_trigger_method_on_autodeploy = await env.get(data.AGENT_TRIGGER_METHOD_ON_AUTO_DEPLOY)
+            push_on_auto_deploy = cast(bool, await env.get(data.PUSH_ON_AUTO_DEPLOY))
+            agent_trigger_method_on_autodeploy = cast(str, await env.get(data.AGENT_TRIGGER_METHOD_ON_AUTO_DEPLOY))
```
[the remark this answers is not in the document]
Comment (wouterdb) [reply src/inmanta/server/services/orchestrationservice.py:393]: isn't the env data also typed in `Environment._settings`  (we could even write a mypy plugin for it ;-) )

[the remark this answers is not in the document]
Comment (bartv) [reply src/inmanta/server/services/orchestrationservice.py:393]: I actually think that the clean way to solve it, is by creating a pydantic subclass with all the settings in there and then all the typing is solved

[the remark this answers is not in the document]
Comment (bartv) [reply src/inmanta/server/services/orchestrationservice.py:393]: Lets keep this for a next round of typing then

**Pull request closed without merging**